kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaAuthnHandler.java

```diff
@@ -0,0 +1,388 @@
+     * <pre><code>
+     *                            START
+     *                              │
+     *       ╭────────────────┬─────┴───────╮───────────────╮
+     *       │                │             │               │
+     *       ↓                ↓             │               ↓
+     * API_VERSIONS ──→ SASL_HANDSHAKE_v0 ══╪══⇒ unframed_SASL_AUTHENTICATE
+     *   │      │                           │                     │
+     *   │      │             ╭─────────────╯                     │
+     *   │      │             ↓                                   │
+     *   │      ╰───→ SASL_HANDSHAKE_v1+ ──→ SASL_AUTHENTICATE    │
+     *   │                                         │              ↓
+     *   ╰─────────────────────────────────────────╰─────→ UPSTREAM_CONNECT
+     * </code></pre>
+     * <ul>
+     * <li>API_VERSIONS if optional within the Kafka protocol</li>
+     * <li>SASL authentication may or may not be in use</li>
+     * </ul>
+     */
+    enum State {
+        START,
+        API_VERSIONS,
+        SASL_HANDSHAKE_v0,
+        SASL_HANDSHAKE_v1_PLUS,
+        UNFRAMED_SASL_AUTHENTICATE,
+        FRAMED_SASL_AUTHENTICATE,
+        AUTHN_SUCCESS
+    }
+
+    public enum SaslMechanism {
+        PLAIN("PLAIN", null),
+        SCRAM_SHA_256("SCRAM-SHA-256",
+                ScramMechanism.SCRAM_SHA_256,
+                SaslInternalConfigs.CREDENTIAL_LIFETIME_MS_SASL_NEGOTIATED_PROPERTY_KEY),
+        SCRAM_SHA_512("SCRAM-SHA-512", ScramMechanism.SCRAM_SHA_512,
+                SaslInternalConfigs.CREDENTIAL_LIFETIME_MS_SASL_NEGOTIATED_PROPERTY_KEY);
+
+        // TODO support OAUTHBEARER, GSSAPI
+        private final String name;
+        private final ScramMechanism scramMechanism;
+        private final String[] negotiableProperties;
+
+        private SaslMechanism(String saslName, ScramMechanism scramMechanism,
+                              String... negotiableProperties) {
+            this.name = saslName;
+            this.scramMechanism = scramMechanism;
+            this.negotiableProperties = negotiableProperties;
+        }
+
+        public String mechanismName() {
+            return name;
+        }
+
+        static SaslMechanism fromMechanismName(String mechanismName) {
+            switch (mechanismName) {
+                case "PLAIN":
+                    return PLAIN;
+                case "SCRAM-SHA-256":
+                    return SCRAM_SHA_256;
+                case "SCRAM-SHA-512":
+                    return SCRAM_SHA_512;
+            }
+            throw new UnsupportedSaslMechanismException(mechanismName);
+        }
+
+        public ScramMechanism scramMechanism() {
+            return scramMechanism;
+        }
+
+        public String[] negotiableProperties() {
+            return negotiableProperties;
+        }
+    }
+
+    // TODO need some way to support SaslAuthenticate v0, which doesn't use Kafka protcol framing at all
```

Comment:
Is this comment still current?  I see you have the BaseSaslR* classes.

Reply:
`BareSaslRequest` and `BareSaslResponse` are in this PR because they're needed to support the original way that Kafka implemented SASL. Prior to KIP-152 clients sent the SASL frames without KRPC encapsulation (within `SaslAuthenticate(Request|Response)`). That isn't a great fit for the existing cases of `(Request|Response)Frame`, which assumes either an `Opaque*Frame` (not decoded at all, therefore the auth bytes are not available to a filter wanting to see them) or `Decoded*Frame` (which is loaded with the assumption there there is a KRP head, with correlation id etc). It seemed simpler to add a third case to `(Request|Response)Frame` than trying to lever this legacy concept into what is otherwise a nice type-safe class hierarchy. 

However, although `BareSaslRequest` and `BareSaslResponse` are in this PR they're not supported by the `KafkaRequestDecoder` etc. So you can see them as a `TODO`.

TBH it's not clear that we really need to implement them at all. I'd expect all recent clients to do the SASL dance properly, but confirming that is a different story.